ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp

```diff
@@ -0,0 +1,206 @@
+    std::vector<std::pair<uint32_t, uint32_t>> padding,
+    const float value,
+    const bool use_multicore,
+    const std::optional<MemoryConfig>& memory_config_arg) {
+
+    const int original_rank = input_tensor.get_shape().rank();
+    if(int diff = original_rank - padding.size(); diff != 0) {
+        TT_FATAL(diff > 0, "ttnn.pad: padding len can't be larger than input tensor rank");
+
+        padding.insert(padding.begin(), diff, {0, 0});
+    }
+
+    TT_FATAL(
+        padding.size() == original_rank,
+        "ttnn.pad: padding must be the same length as the input tensor rank");
+    TT_FATAL(
+        input_tensor.get_layout() != ttnn::ROW_MAJOR_LAYOUT,
+        "ttnn.pad: row-major tensors have to use fallback because the kernel currently causes a PCC error");
+
+    // Unsqueeze Tensor to 4D if it is not already
+    ttnn::Tensor input_tensor_4D = ttnn::unsqueeze_to_4D(input_tensor);
+
+
+    padding.insert(padding.begin(), 4 - original_rank, {0, 0});
+    auto input_shape_with_tile_padding = input_tensor_4D.get_shape().with_tile_padding();
+
+    ShapeType output_padded_shape;
+    for(size_t i = 0; i < padding.size(); i++) {
+        output_padded_shape[i] = input_shape_with_tile_padding[i] + padding[i].second;
+    }
+
+    auto pad_front = padding | std::views::transform([](const auto& p) { return p.first; });
+    auto pad_back = padding | std::views::transform([](const auto& p) { return p.second; });
+
+    const bool front_padding_is_zero = std::accumulate(pad_front.begin(), pad_front.end(), 0) == 0;
+    TT_FATAL(
+        front_padding_is_zero,
+        "ttnn.pad: on device padding does not support front padding");
+
+    const int target_height = output_padded_shape[padding.size() - 2];
+    const int target_width = output_padded_shape[padding.size() - 1];
+    TT_FATAL(
+        target_height % ttnn::TILE_SIZE == 0 || target_width % ttnn::TILE_SIZE == 0,
+        "ttnn.pad: for tiled tensors padding end must be a multiple of the tile size on height and width for a "
+        "tensor in tile layout");
+
+    // Performing actual padding
+    ShapeType pad_front_array;
+    for(size_t i = 0; i < pad_front.size(); i++) {
+        pad_front_array[i] = pad_front[i];
+    }
+
+    return pad_impl<ShapeType>(queue_id, input_tensor_4D, output_padded_shape, pad_front_array, value, use_multicore, memory_config_arg);
+}
+
+} // anonymous namespace
+
+// This function signature is similar to pytorch's signature
+// Any rank tensor supported
+static ttnn::Tensor operator()(
+    uint8_t queue_id,
+    const ttnn::Tensor& input_tensor,
+    std::vector<std::pair<uint32_t, uint32_t>> padding,
+    const float value,
+    const bool use_multicore,
+    const std::optional<MemoryConfig>& memory_config_arg) {
+    const int original_rank = input_tensor.get_shape().rank();
+
+    ttnn::Tensor output_tensor;
+    if (input_tensor.storage_type() != StorageType::DEVICE) {
+        if(original_rank == 1) {
```

Comment:
I tried to simplify this, but did not succeed. Left as is.

Reply:
switch (original_rank) {
            case 1: output_tensor = pad_impl<tt::tt_metal::Array1D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 2: output_tensor = pad_impl<tt::tt_metal::Array2D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 3: output_tensor = pad_impl<tt::tt_metal::Array3D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 4: output_tensor = pad_impl<tt::tt_metal::Array4D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 5: output_tensor = pad_impl<tt::tt_metal::Array5D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 6: output_tensor = pad_impl<tt::tt_metal::Array6D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 7: output_tensor = pad_impl<tt::tt_metal::Array7D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 8: output_tensor = pad_impl<tt::tt_metal::Array8D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            default: TT_FATAL("Unsupported tensor rank");
        }

Reply:
```
switch (original_rank) {
            case 1: output_tensor = pad_impl<tt::tt_metal::Array1D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 2: output_tensor = pad_impl<tt::tt_metal::Array2D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 3: output_tensor = pad_impl<tt::tt_metal::Array3D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 4: output_tensor = pad_impl<tt::tt_metal::Array4D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 5: output_tensor = pad_impl<tt::tt_metal::Array5D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 6: output_tensor = pad_impl<tt::tt_metal::Array6D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 7: output_tensor = pad_impl<tt::tt_metal::Array7D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            case 8: output_tensor = pad_impl<tt::tt_metal::Array8D>(queue_id, input_tensor, padding, value, use_multicore, memory_config_arg); break;
            default: TT_FATAL("Unsupported tensor rank");
        }
```

Reply:
oh, yes, that should work